### ISSUE TITLE
<%= taxonomy %> doesn't work in README.md.erb

### ISSUE THREAD

**Issue by DavidGoldwasser** [NREL/OpenStudio#3156]
https://github.com/NREL/OpenStudio/blob/6aed934ea899f50ee017347bf4333687f5af15ca/openstudiocore/src/utilities/bcl/templates/common/README.md.erb#L17

This isn't a big deal, but if it is an easy fix and if we have to make a new installer anyway, it would be nice to fix. This will the readme.md on the all the updated measures on BCL.

Note: this doesn't error out the process, it just ends up with a blank space on the Readme.md under taxonomy.

**Comment by macumber** [comment]
Looks like the issue here is that I added support for putting taxonomy in the measure.rb which will update it in the XML.  Taxonomy in measure.rb will end up in the readme.md but not if it is only in the xml.

  def taxonomy
    return 'Envelope.form'
  end

**Comment by DavidGoldwasser** [comment]
Ok, so more of a measure update issue than core C++. That makes sense. I can close this issue unless you see reason to keep it open. For measures, I think would probably just add the def to measure.rb as measure are worked on. I don't think we have the time or an automated way to batch change them.

**Comment by DavidGoldwasser** [comment]
I'm going to re-open this as it is something I want to show on all measure.rb files to identify where the users will find these measures on BCL.

**Comment by jmarrec** [comment]
I am not following what the action item(s) is (are) @DavidGoldwasser? Could you explain please?

If I follow correctly:
* Taxonomy is more meant to be inside the measure.xml as `tag` https://nrel.github.io/OpenStudio-user-documentation/reference/measure_writing_guide/#measure-metadata-file-measurexml
* But it was implemented as a method `def taxonomy` inside the measure.rb

So how do we resolve this?

**Comment by jmarrec** [comment]
I'm not sure why the XML would allow several tags but setTaxonomyTag is single-use only.

https://github.com/NREL/OpenStudio/blob/develop/src/utilities/bcl/BCLMeasure.hpp#L190-L195

**Comment by DavidGoldwasser** [comment]
@jmarrec I wan to pull @nllong and @kflemin in on this. While the SDK may allow multiple, I believe for BCL, and I think for the OpenStudio Application, it was used less of a tag and more fo a classification, with every measure having exactly one taxonomy. 

My reason for wanting to show this in the readme.md file is so that a user who finds a measure on GitHub and wants to find it on BCL or in the Application doesn't have to dig around the measure.xml file on their own. It is not always obvious where it should be. For example there is no proper place for measure that deal with site like change building location, it is in Whole Building.Space Types which has been a landing spot for lost measure since it makes more sense the the default of Envelope.Form when you make a new measure.

**Comment by DavidGoldwasser** [comment]
@jmarrec I understand the confusion about if there is a single or multiple tags, but is that related to the problems I'm having getting a tag or tags to show in the readme.md file? Can we solve one without solving the other?

**Comment by macumber** [comment]
I think the issue would be in:

https://github.com/NREL/OpenStudio/blob/e603410ac3567d44e7660294f3b06a8a2d591118/src/cli/measure_manager.rb#L312
https://github.com/NREL/OpenStudio/blob/e603410ac3567d44e7660294f3b06a8a2d591118/src/cli/measure_manager.rb#L501

I don't know if the new BCL supports multiple tags, the single taxonomy, or both?

**Comment by jmarrec** [comment]
@DavidGoldwasser There really isn't much of an issue currently if we allow just one tag. The measure template should just be updated to have a placeholder for it (and perhaps it should be documented a bit better *somewhere*).

In your `measure.rb` add the following:

```ruby
  # A list of BCL tags https://bcl.nrel.gov/api/taxonomy/measure
  def taxonomy
    return 'Envelope.Form'
  end
```

Run `openstudio measure -u .` => You'll see it in **both** the `measure.xml` and `measure.rb`

Now, do this change in `measure.rb`

```diff
  def taxonomy
-    return 'Envelope.Form'
+    return 'HVAC.HVAC Controls'
  end
```
 
Run `openstudio measure -u .` => You'll see it updated just fine in **both** the `measure.xml` and `measure.rb` as well

**Comment by jmarrec** [comment]
> Looks like the issue here is that I added support for putting taxonomy in the measure.rb which will update it in the XML. Taxonomy in measure.rb will end up in the readme.md but not if it is only in the xml.
> 
> def taxonomy
> return 'Envelope.form'
> end

**Comment by DavidGoldwasser** [comment]
@jmarrec sorry I missed a few things, I didn't know I could add taxonomy to measure.rb none of our measures have that. I also was hitting error fo r`"error":"undefined method `each' for \"Envelope.Form\` with your code, but can just use what I already have in the readme.md.erb if I make the new addition to measure.rb.

I'll have to update every measure that I want to use this on, but I guess that is also true for bring over description and modeler description from the measure.xml file.

**Comment by jmarrec** [comment]
Yeah that error means you probably tried to make it a list but didn't update the readme.md.erb to reflect that.

I think we still need to clean up our act here so I'll reopen:

* [ ] Question: do we keep supporting a single taxonomy tag? @kflemin @nllong 
* [ ] Based on the outcome of that question, update the measure templates to include the `def taxonomy` method.
* [ ] (maybe some updates to the doc *somewhere*?)

**Comment by nllong** [comment]
On the question--i think that is a question on how the new BCL is handing it. Also need to make sure that PAT handles it too (may be moot if using the MeasureManager). @kflemin

**Comment by jmarrec** [comment]
> @kflemin @nllong is the new BCL ready or do we have design docs to answer this question please? I'd like to put this issue to rest as soon as possible so we can move on. Thanks!

Does the new BCL handle multiple Taxonomy tags?

**Comment by jmarrec** [comment]
Ping

**Comment by DavidGoldwasser** [comment]
I can try to close this or get fix in for 3.11 if there is something else. I'll tag it so I don't forget about it

**Comment by jmarrec** [comment]
Ping

**Comment by DavidGoldwasser** [comment]
I think we will have some new XML data soon related to measure language and possibly additional dependences. We we update to add those to measure.rb we can fix this as well. So I may tag this for 3.11.0 instead of 3.10.0. @kflemin does that make sense.
